generator/geometry_holder.hpp

```diff
@@ -58,14 +58,20 @@ class GeometryHolder
 
   Points const & GetSourcePoints()
   {
+    // For short lines keep simplifying the previous version to ensure points visibility is consistent.
     return !m_current.empty() ? m_current : m_fb.GetOuterGeometry();
   }
 
+  // Its important AddPoints is called sequentially from upper scales to lower.
   void AddPoints(Points const & points, int scaleIndex)
   {
     if (m_ptsInner && points.size() <= m_maxNumTriangles)
     {
+      // Store small features inline and keep a mask for individual points scale visibility.
       if (m_buffer.m_innerPts.empty())
+        // FIXME: if geometry is added for the most detailed scale 3 only
+        // then the mask is never updated and left == 0,
+        // which means the geometry could be used on any scale.
```

Comment:
This comment is controversial. Mask is not needed, because a feature is invisible on other scales if AddPoints was not called.

Reply:
It matters for style testing/design cases only.

E.g. if line feature's visibility is changed from z15 to z14, then the lines won't appear on z14 (even after visibility/scale index regen), because there is no geometry - its kinda expected.
But actually some of the lines do appear! But degraded to straight lines always.
And this is confusing and ugly.

So its not a big issue indeed but nice (and easy) to fix still.

Reply:
> This comment is controversial. Mask is not needed, because a feature is invisible on other scales if AddPoints was not called.

But you're right it doesn't need any change if this patch is accepted, because for style designers these mask==0 points will be used as a fallback geometry anyway.